master/internal/api_project.go

```diff
@@ -197,72 +197,61 @@ func (a *apiServer) getProjectColumnsByID(
 		}
 	}
 	summaryMetrics := []struct {
-		ID             int
-		SummaryMetrics model.JSONObj
+		MetricName string
+		JSONPath   string
+		MetricType string
+		Count      int32
 	}{}
 
 	if len(trialIDs) > 0 {
-		trialsQuery := db.Bun().NewSelect().
-			Column("id", "summary_metrics").
-			Table("trials").
-			Where("id IN (?)", bun.In(trialIDs)).
-			Order("id")
+		subQuery := db.BunSelectMetricGroupNames().ColumnExpr(
```

Comment:
Query plan attached
`                                                                                                                                                                              QUERY PLAN                                                                   
                                                                                                            
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------
 WindowAgg  (cost=20776280.83..20776303.33 rows=1000 width=104) (actual time=712.964..713.194 rows=41 loops=1)
   Output: stats.json_path, stats.metric_name, stats.metric_type, row_number() OVER (?)
   CTE stats
     ->  Sort  (cost=20776208.50..20776211.00 rows=1000 width=96) (actual time=712.796..712.803 rows=41 loops=1)
           Output: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))))), ((((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))) -> (jsonb_object_k
eys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))) ->> 'type'::text))
           Sort Key: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))
           Sort Method: quicksort  Memory: 28kB
           ->  HashAggregate  (cost=20776121.17..20776158.67 rows=1000 width=96) (actual time=712.610..712.657 rows=41 loops=1)
                 Output: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))))), ((((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))) -> (jsonb_ob
ject_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))) ->> 'type'::text))
                 Group Key: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))))), (((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))) -> (jsonb_
object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))) ->> 'type'::text)
                 ->  Result  (cost=0.00..17672546.17 rows=413810000 width=96) (actual time=8.410..712.504 rows=314 loops=1)
                       Output: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))))), (((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))) -> (jso
nb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))) ->> 'type'::text)
                       ->  ProjectSet  (cost=0.00..2154671.17 rows=413810000 width=69) (actual time=8.408..711.782 rows=314 loops=1)
                             Output: jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))), trials.summary_metrics, (jsonb_object_keys(trials.summary_metrics))
                             ->  ProjectSet  (cost=0.00..23549.67 rows=4138100 width=37) (actual time=8.406..711.537 rows=155 loops=1)
                                   Output: jsonb_object_keys(trials.summary_metrics), trials.summary_metrics
                                   ->  Seq Scan on public.trials  (cost=0.00..2548.81 rows=41381 width=5) (actual time=8.396..698.490 rows=41415 loops=1)
                                         Output: trials.summary_metrics
                                         Filter: (trials.summary_metrics IS NOT NULL)
   ->  Sort  (cost=69.83..72.33 rows=1000 width=96) (actual time=712.950..712.952 rows=41 loops=1)
         Output: stats.json_path, stats.metric_name, stats.metric_type
         Sort Key: stats.json_path, stats.metric_name, stats.metric_type
         Sort Method: quicksort  Memory: 28kB
         ->  CTE Scan on stats  (cost=0.00..20.00 rows=1000 width=96) (actual time=712.801..712.882 rows=41 loops=1)
               Output: stats.json_path, stats.metric_name, stats.metric_type
 Planning time: 10.639 ms
 Execution time: 713.969 ms
(27 rows)`

This is done on 4.1k trials without any filter on trial id, but I think the sequential scan is fine, most of the time is spent at breaking the summary metrics json. Is this acceptable or should we create a cache to aggregate by project id?

Reply:
i dont know how frequently web calls `GetProjectColumns` but adding 1s more to this endpoint seems like a non-starter to me. maybe a 2nd web opinion is useful?

Reply:
seq scan of trials outside of a migration or background agg seems like a bad idea.

Reply:
Right, a project with couple experiments would look like this
```
                                                                                                                                                                             QUERY PLAN                                                                                     
                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------
 WindowAgg  (cost=11363.16..11383.41 rows=900 width=104) (actual time=1.641..1.655 rows=18 loops=1)
   CTE stats
     ->  Sort  (cost=11298.75..11301.00 rows=900 width=96) (actual time=1.526..1.528 rows=18 loops=1)
           Sort Key: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))
           Sort Method: quicksort  Memory: 26kB
           ->  HashAggregate  (cost=11220.84..11254.59 rows=900 width=96) (actual time=1.456..1.464 rows=18 loops=1)
                 Group Key: (jsonb_object_keys(trials.summary_metrics)), (jsonb_object_keys((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))))), (((trials.summary_metrics -> (jsonb_object_keys(trials.summary_metrics))) -> (jsonb_object_keys((tria
ls.summary_metrics -> (jsonb_object_keys(trials.summary_metrics)))))) ->> 'type'::text)
                 ->  Result  (cost=94.55..9570.84 rows=220000 width=96) (actual time=0.227..1.387 rows=186 loops=1)
                       ->  ProjectSet  (cost=94.55..1320.84 rows=220000 width=69) (actual time=0.210..0.715 rows=186 loops=1)
                             ->  ProjectSet  (cost=94.55..187.84 rows=2200 width=37) (actual time=0.176..0.419 rows=74 loops=1)
                                   ->  Bitmap Heap Scan on trials  (cost=94.55..176.68 rows=22 width=5) (actual time=0.123..0.267 rows=22 loops=1)
                                         Recheck Cond: (id = ANY ('{41693,41688,41686,41648,41647,41646,41644,41640,41598,1074,1073,1070,1069,1076,614,41694,41692,41691,41690,41689,41687,41645}'::integer[]))
                                         Filter: (summary_metrics IS NOT NULL)
                                         Heap Blocks: exact=17
                                         ->  Bitmap Index Scan on trials_pkey  (cost=0.00..94.55 rows=22 width=0) (actual time=0.103..0.103 rows=22 loops=1)
                                               Index Cond: (id = ANY ('{41693,41688,41686,41648,41647,41646,41644,41640,41598,1074,1073,1070,1069,1076,614,41694,41692,41691,41690,41689,41687,41645}'::integer[]))
   ->  Sort  (cost=62.16..64.41 rows=900 width=96) (actual time=1.619..1.621 rows=18 loops=1)
         Sort Key: stats.json_path, stats.metric_name, stats.metric_type
         Sort Method: quicksort  Memory: 26kB
         ->  CTE Scan on stats  (cost=0.00..18.00 rows=900 width=96) (actual time=1.528..1.532 rows=18 loops=1)
 Planning time: 15.668 ms
 Execution time: 1.935 ms
(22 rows)
```

Reply:
Upon discussing with @ashtonG , we want to restore the deduplication logic from database level to go code level, and this would keep everything as performant as right now.

Reply:
either way is fine with me, i didn't realize we were already doing it in go code, didn't have the history.

i could go for `DISTINCT` in postgres but the aggregation in Go as being easiest to read. i can't say which is faster. probably all in postgres, if i had to guess.